documentation/projects/proposals/dark_mode/20240325-implementation_plan_dark_mode.md

```diff
@@ -0,0 +1,326 @@
+
+         1. Replace the "hardcoded" color values in the Tailwind configuration
+            file with css variables defined in the "base" layer of the
+            `tailwind.css` file. This will not result in any visual changes.
+
+      2. Add the dark mode colors as CSS variable definitions, nested under the
+         `.dark-mode` CSS class, in the "base" layer of the `tailwind.css` file.
+         This will not result in any visual changes, _except when
+         `FORCE_DARK_MODE` is enabled._
+
+      3. Visual Regression tests. Update
+         `frontend/test/playwright/utils/breakpoints.ts` so that each breakpoint
+         produces and expects a dark mode screenshot to pass as well as the
+         existing light mode screenshot. **This will also be a significant diff,
+         as at the time of writing it will create 293 new screenshots to review.
+         This is also the point of the process where @fcoveram and
+         @wordpress/openverse-frontend should review the full dark mode
+         appearance for correctness and sufficient color contrast (see the
+         ["Accessibility"](#accessibility) section for more details.** Existing
+         screenshots should be renamed to `<snapshot_name>_light` and the new
+         dark mode screenshots should be named `<snapshot_name>_dark`.
+
+2. **Work Stream B**: Toggling dark mode
+
+   1. Create a `DARK_MODE_UI_TOGGLE` feature flag which is off by default and
+      switchable in staging.
+   2. Implement logic for calculating a current "color mode" with the following
+      state:
+
+      ```ts
+      interface ColorMode {
+        preference: "dark" | "light" | "system" // Defaults to "system"
+        systemValue: "light" | "dark" // Readonly representation of the system value
+      }
+      ```
+
+      The color mode should be stored in a cookie so that a previously-selected
+      user choice can be used when rendering via SSR and prevent a visual flash
+      of the incorrect color mode. The "system" value is the default.
+
+   3. Behind the feature flag, add the new user interface element which toggles
+      dark mode (exact design TBD, but it will be comprised of existing UI
+      components). Supports choosing between "dark", "light", and "system"
+      modes, with "system" as the default. The "dark" and "light" options will
+      set a `.dark-mode` or `.light-mode` class on the HTML element of the site.
+      The default "system" choice does not add a class to the HTML element. When
+      there is no HTML `.{color}-mode` class present, the site will default to
+      the `prefers-color-scheme` media query value.
+      1. Add a `TOGGLE_COLOR_SCHEME` analytics event with a playload including
+         the color mode preference chosen by the user.
+      2. Update our Cloudflare static page caching rule for the frontend in with
+         a Cookie bypass rule (in pseudocode, something like `and not
+         http.cookie contains "openverse_color_scheme"))
```

Comment:
Would this make all of the users who select the dark mode bypass cache?

Reply:
@obulat yes, it would, but crucially only on the pages which are being cached: the homepage and all static pages. Also, due to SPA routing these, static pages are only served from the cache on the user's **initial page load**. Therefore, cached homepage and static page visits mark an overall small percentage of Openverse traffic.

Reply:
Just wanted to add more 'flavour' to the issue, but even if we were worried about the caching issue, we could solve it by caching and segmenting the cache at the nginx level (or replace it with varnish, something more suitable for complex caching rules). We'd already be doing that if we moved those routes "statically" out of the app or cached them _at all_, and we could segment on as many relevant cookies as needed (hopefully not too many!).

Just say, there are solutions to this if it ever became a problem.